### PR TITLE
Extract nextgen

### PR DESCRIPTION
**Pull request by jhennawi**
Hi All,

This is the update of the previous extraction PR that I submitted earlier, but then closed. That was because arclines was preventing it from clearing travis, which should now be fixed after @rcooke-ast PR.

I implemented most of the feedback you guys gave me before, and have tightened up many other things by running on the dev suite.

Currently all the valid instruments are passing on the dev suite.

### CODE REVIEW

**Comment by jhennawi** [comment]
Hi @profxj and @rcooke-ast, 

I'm not sure why travis is not clearing here. The tests that are failing here are test_arc.py and test_arclines.py.  

These routines have nothing to do with any code I've developed. Furthermore they are passing fine on my machine. I'm not sure what the issue is here, but I think @rcooke-ast had a similar problem. Since I've pulled in from master, in principle this should be fixed so I'm a bit lost. 

Joe

File: CHANGES.rst [hunk 1]
```diff
@@ -9,6 +9,8 @@
 - Turn off 2.7 Travis testing
 - Integrated arclines into PypeIt
 - Modified debug/developer modes
+- Update SpecObjs class; ndarray instead of list;  set() method
```
Comment (profxj) [inline CHANGES.rst]: I should think there were *many* more CHANGES than this.

File: pypeit/core/save.py [hunk 2]
```diff
@@ -399,47 +399,49 @@ def save_1d_spectra_fits(specobjs, header, outfile, helio_dict=None, telescope=N
     # Loop on detectors
     npix = 0
     ext = 0
-    for specobj in all_specobj:
-        if specobj is None:
+    for spec in specobjs:
```
Comment (profxj) [inline pypeit/core/save.py]: If ```specobjs``` is a ```SpecObjs``` object as the docs state, then I feel this should
be

```for spec in specobjs.specbojs```

I suspect the docs are wrong.

File: pypeit/core/save.py [hunk 3]
```diff
@@ -399,47 +399,49 @@ def save_1d_spectra_fits(specobjs, header, outfile, helio_dict=None, telescope=N
...
+        if spec is None:
```
Comment (profxj) [inline pypeit/core/save.py]: Using ```spec``` here is a bit funny.  I should prefer ```sobj```
But I'll live with it as is.

File: pypeit/core/skysub.py [hunk 4]
```diff
@@ -141,7 +141,8 @@ def bg_subtraction_slit(slit, slitpix, edge_mask, sciframe, varframe, tilts,
 
 
 # ToDO Fix masking logic. This code should also take an ivar for consistency with rest of extraction
-def global_skysub(image, ivar, tilts, thismask, slit_left, slit_righ, inmask = None, bsp=0.6, sigrej=3., TRIM_EDG = (3,3), POS_MASK=True, PLOT_FIT=False):
+def global_skysub(image, ivar, tilts, thismask, slit_left, slit_righ, inmask = None, bsp=0.6, sigrej=3., TRIM_EDG = (3,3),
```
Comment (profxj) [inline pypeit/core/skysub.py]: Is SHOW_FIT in the doc string?

File: pypeit/core/skysub.py [hunk 5]
```diff
@@ -607,18 +604,45 @@ def local_skysub_extract(sciimg, sciivar, tilts, waveimg, global_sky, rn2_img, t
...
-                color = 'green'
+            if spec.HAND_EXTRACT_FLAG is False:
+                color = 'magenta'
             else:
                 color = 'orange'
             ginga.show_trace(viewer, ch, spec.trace_spat, spec.idx, color=color)
-        # TODO figure out a way to set the cuts of the ginga viewer to go from -5 to 5
+
+        # These are the pixels that were masked by the extraction
+        spec_mask, spat_mask = np.where((outmask == False) & (inmask == True))
```
Comment (profxj) [inline pypeit/core/skysub.py]: should be ```is``` and not ```=```
for both.  or better,   

```np.where(~outmask & inmask)```

File: pypeit/par/pypeitpar.py [hunk 6]
```diff
@@ -1448,7 +1530,7 @@ def validate(self):
             self.data['params'] = [self.data['params']]
         pass
 
-
+# TODO JFH This parset is now deprecated.
```
Comment (profxj) [inline pypeit/par/pypeitpar.py]: Why keep the deprecated parset's ?

File: pypeit/pypeit.py [hunk 7]
```diff
@@ -90,6 +93,13 @@ def PypeIt(pypeit_file, setup_only=False, calibration_check=False, use_header_fr
                                                        calibration_check=calibration_check,
                                                        use_header_frametype=use_header_frametype,
                                                        sort_dir=sort_dir)
+
+    # TODO This --use_masters flag should probably be passed as an argument to setup instead of modifying the parset here
+    # If the --use_masters flag was set change the parset to make this so
+    if(use_masters):
```
Comment (profxj) [inline pypeit/pypeit.py]: no need for ```()```

File: pypeit/scripts/view_fits.py [hunk 8]
```diff
@@ -57,42 +57,46 @@ def main(args):
...
+#        hdulist.writeto(kludge_fil,clobber=True)
         #
-        args.file = kludge_fil
+#        args.file = kludge_fil
 
     # RAW_LRIS??
     if args.raw_lris:
         # 
         img, head, _ = keck_lris.read_lris(args.file)
+        ginga.show_image(img)
```
Comment (profxj) [inline pypeit/scripts/view_fits.py]: Am glad you've eliminated this kludge.
Might as well eliminate the commented lines too.

File: pypeit/utils.py [hunk 9]
```diff
@@ -237,7 +239,7 @@ def bspline_profile(xdata, ydata, invvar, profile_basis, upper=5, lower=5,
             # we'll do the fit right here..............
             if error != 0:
                 bf1, laction, uaction = sset.action(xdata)
-                if(bf1.size !=nx*nord):
+                if((bf1 is -2) or (bf1.size !=nx*nord)):
```
Comment (profxj) [inline pypeit/utils.py]: ```bf1 == -2 ```

File: pypeit/core/skysub.py (hunk 5 above)
Comment (jhennawi) [reply pypeit/core/skysub.py]: You are confused here @profxj.  is and not are logical operators but they are *not* bitwise operators. outmask is False will evaluate to a single boolean and not a numpy boolean array.

Comment (jhennawi) [reply pypeit/core/skysub.py]: Your second example is correct though.

File: CHANGES.rst (hunk 1 above)
Comment (jhennawi) [reply CHANGES.rst]: More changes added.

File: pypeit/core/save.py (hunk 2 above)
Comment (jhennawi) [reply pypeit/core/save.py]: Fixed.

File: pypeit/utils.py (hunk 9 above)
Comment (jhennawi) [reply pypeit/utils.py]: Thanks! Fixed.

File: pypeit/scripts/view_fits.py (hunk 8 above)
Comment (jhennawi) [reply pypeit/scripts/view_fits.py]: Fixed!

File: pypeit/pypeit.py (hunk 7 above)
Comment (jhennawi) [reply pypeit/pypeit.py]: Fixed.

File: pypeit/par/pypeitpar.py (hunk 6 above)
Comment (jhennawi) [reply pypeit/par/pypeitpar.py]: I may need them as code examples.

File: pypeit/core/save.py (hunk 3 above)
Comment (jhennawi) [reply pypeit/core/save.py]: Fixed to sobj.

File: pypeit/core/skysub.py (hunk 4 above)
Comment (jhennawi) [reply pypeit/core/skysub.py]: Doc string updated.

**Pull request merged**